### PR TITLE
feat: Add findByImage support to Windows driver

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
https://github.com/appium/appium/issues/14073

### CODE REVIEW

**Comment by jsa34** [comment]
Wow - such a quick response!  Thank you!  This is a big blocker for me at the moment with what I am trying to do, so much appreciated!

**Comment by mykola-mokhnach** [comment]
@jsa34 Could you please check out this code and test it locally? With your help it would go faster

**Comment by jsa34** [comment]
Of course!  Let me see...

**Comment by jsa34** [comment]
Trying by npm is being difficult, won't let me install git repo package, just stuck at rollbackFailedOptional...

**Comment by mykola-mokhnach** [comment]
You can simply install appium beta and then replace or link (see `npm link`) the original windows driver with the checked out content. Also make sure the package is transpiled before injecting it (running `npm install` should do the magic)

**Comment by jsa34** [comment]
Probably just me as I am much more of a Java kind of guy, but using windows here and npm link just keeps giving me "Access is denied" response, but nothing seems weird with permissions that I can see...

**Comment by mykola-mokhnach** [comment]
https://github.com/appium/appium/issues/13227#issuecomment-531812534 contains more detailed tutorial. Simply replace `webdriveragent-repo` with the windows driver's one

**Comment by jsa34** [comment]
Still seem to get the same response, but think I've followed the above guide correctly:

Error in IDE:
Caused by: org.openqa.selenium.InvalidSelectorException: Invalid selector: -image (WARNING: The server did not provide any stacktrace information)

Appium:
[HTTP] --> POST /wd/hub/session/8fccee01-f83f-4bdb-8d81-0a78453fbea7/element
[HTTP] {"using":"-image","value":"***"}
[MJSONWP (8fccee01)] Driver proxy active, passing request on via HTTP proxy
[debug] [WD Proxy] Matched '/wd/hub/session/8fccee01-f83f-4bdb-8d81-0a78453fbea7/element' to command name 'findElement'
[debug] [WD Proxy] Proxying [POST /wd/hub/session/8fccee01-f83f-4bdb-8d81-0a78453fbea7/element] to [POST http://127.0.0.1:4724/wd/hub/session/DD258CF1-CE6A-4875-80BA-56D9BA28F94A/element] with body: {"using":"-image","value":"***"}
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] ==========================================
[WinAppDriver] [STDOUT] POST /wd/hub/session/DD258CF1-CE6A-4875-80BA-56D9BA28F94A/element HTTP/1.1
[WinAppDriver] [STDOUT] Accept: application/json, */*
[WinAppDriver] [STDOUT] Connection: close
[WinAppDriver] [STDOUT] Content-Length: 737
[WinAppDriver] [STDOUT] Content-Type: application/json; charset=utf-8
[WinAppDriver] [STDOUT] Host: 127.0.0.1:4724
[WinAppDriver] [STDOUT] User-Agent: appium
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] {"using":"-image","value":"***"}
[WinAppDriver] [STDOUT] HTTP/1.1 400 Bad Request
[WinAppDriver] [STDOUT] Content-Length: 87
[WinAppDriver] [STDOUT] Content-Type: application/json
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] {"status":32,"value":{"error":"invalid selector","message":"Invalid selector: -image"}}
[WD Proxy] Got response with status 400: {"status":32,"value":{"error":"invalid selector","message":"Invalid selector: -image"}}
[debug] [MJSONWP (8fccee01)] Encountered internal error running command: ProxyRequestError: Could not proxy command to remote server. Original error: 400 - {"status":32,"value":{"error":"invalid selector","message":"Invalid selector: -image"}}
[debug] [MJSONWP (8fccee01)]     at JWProxy.proxy (C:\Users\automation\AppData\Roaming\npm\node_modules\appium-windows-driver\node_modules\appium-base-driver\lib\jsonwp-proxy\proxy.js:218:13)
[debug] [MJSONWP (8fccee01)]     at runMicrotasks (<anonymous>)
[debug] [MJSONWP (8fccee01)]     at runNextTicks (internal/process/task_queues.js:62:5)
[debug] [MJSONWP (8fccee01)]     at processImmediate (internal/timers.js:429:9)
[debug] [W3C] Matched W3C error code 'invalid selector' to InvalidSelectorError
[HTTP] <-- POST /wd/hub/session/8fccee01-f83f-4bdb-8d81-0a78453fbea7/element 500 33 ms - 111

**Comment by mykola-mokhnach** [comment]
> Still seem to get the same response, but think I've followed the above guide correctly:

This is easy to check, simply add something like `throw new Error('yolo');` to the `findElOrEls` method to see whether it is called or not ;)

**Comment by jsa34** [comment]
Yep, as expected, it isn't calling find.js.

I had noticed this before, though, and it may be the root cause of the problem: No matter what I do, all .findElement calls, etc, return back an element of type WebElement, and I can never get a WindowsElement.  In the error stacktrace, it looks like appium is resolving to the native selenium find methods so never looks for find.js...
Caused by: org.openqa.selenium.InvalidSelectorException: Invalid selector: -image (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds
For documentation on this error, please visit: https://www.seleniumhq.org/exceptions/invalid_selector_exception.html
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: '***', ip: '***', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_221'
Driver info: io.appium.java_client.windows.WindowsDriver
Capabilities {app: Root, deviceName: WindowsPC, experimental-webdriver: true, javascriptEnabled: true, platform: WINDOWS, platformName: Windows}
Session ID: 2afdf5b7-ca93-4ce7-9e20-39bc4ddf04e8
*** Element info: {Using=-image, value=***}
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:214)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:166)
	at org.openqa.selenium.remote.http.JsonHttpResponseCodec.reconstructValue(JsonHttpResponseCodec.java:40)
	at org.openqa.selenium.remote.http.AbstractHttpResponseCodec.decode(AbstractHttpResponseCodec.java:80)
	at org.openqa.selenium.remote.http.AbstractHttpResponseCodec.decode(AbstractHttpResponseCodec.java:44)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
	at io.appium.java_client.remote.AppiumCommandExecutor.execute(AppiumCommandExecutor.java:239)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:552)
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:41)
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
	at io.appium.java_client.windows.WindowsDriver.execute(WindowsDriver.java:1)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:323)
	at io.appium.java_client.DefaultGenericMobileDriver.findElement(DefaultGenericMobileDriver.java:61)
	at io.appium.java_client.AppiumDriver.findElement(AppiumDriver.java:1)
	at io.appium.java_client.windows.WindowsDriver.findElement(WindowsDriver.java:1)
	at io.appium.java_client.FindsByImage.findElementByImage(FindsByImage.java:42)
	at io.appium.java_client.MobileBy$ByImage.findElement(MobileBy.java:567)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:315)
	at io.appium.java_client.DefaultGenericMobileDriver.findElement(DefaultGenericMobileDriver.java:57)
	at io.appium.java_client.AppiumDriver.findElement(AppiumDriver.java:1)
	at io.appium.java_client.windows.WindowsDriver.findElement(WindowsDriver.java:1)
	at org.openqa.selenium.support.ui.ExpectedConditions$6.apply(ExpectedConditions.java:182)
	at org.openqa.selenium.support.ui.ExpectedConditions$6.apply(ExpectedConditions.java:179)
	at io.appium.java_client.AppiumFluentWait.until(AppiumFluentWait.java:222)

**Comment by jsa34** [comment]
Just trying with a simple driver.findElementByImage results in the same, all the findElementBy* return WebElements instead of WindowsElements for the following simple code (Tried to make sure at each step drivers are definitely casted to WindowsDriver and elements are casted to WindowsElements if some reason they aren't) ...

```
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability("deviceName", "WindowsPC");
        capabilities.setCapability("app", "Root");
        capabilities.setCapability("ms:experimental-webdriver", true);

        WindowsDriver<WindowsElement> driver = new WindowsDriver<>(currentDriver.getRemoteAddress(), capabilities);
        WindowsElement windowsElement =  (WindowsElement) driver.findElementByImage(imageName);

```

**Comment by mykola-mokhnach** [comment]
Are you sure you use appium@beta as the base? With 1.16 it won't work for sure

**Comment by jsa34** [comment]
Yes:

[Appium] Welcome to Appium v1.17.0-beta.1 (REV 8cffb5274854c4cc1e639d56fb2e539c6a5c1786)
[Appium] Appium REST http interface listener started on 0.0.0.0:4723

Is logged on startup

**Comment by mykola-mokhnach** [comment]
ok, this is fine. Now it might necessary to check if the endpoint is proxied properly. For that you might add some logging to the appium-base-driver module:
- https://github.com/appium/appium-base-driver/blob/8624a8061f836c14e331aeace39a352d31fcb875/lib/basedriver/driver.js#L491 to make sure the `/element` endpoint does not get proxied immediately
- https://github.com/appium/appium-base-driver/blob/8624a8061f836c14e331aeace39a352d31fcb875/lib/basedriver/commands/find.js#L47 to make sure the base driver's findElement is called first

**Comment by mykola-mokhnach** [comment]
the simplest way to add logging would be to just modify the corresponding files inside `build` folder of the corresponding module. `console.log('>>> my log');` would be enough to see the result in appium log

File: lib/commands/find.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import _ from 'lodash';
+import { errors } from 'appium-base-driver';
+import { util } from 'appium-support';
+
+
+const commands = {};
+
+commands.findElOrEls = async function findElOrEls (strategy, selector, mult, context) {
```
Comment (jlipps) [inline lib/commands/find.js]: seems like we could have a generic `findElOrElsByProxy` method since we do this in many drivers.?

**Comment by jsa34** [comment]
Managed to find the problem - it was silently failing to update the build after doing an install... Fixed now.

So, I have 2 issues: 
1. In the current code, previous requests for locator strategies such as By.name, By.xpath are now throwing invalid locator errors. However, locator -image is fine!  It too throws an error but saying:
`[BaseDriver] Finding image element with match threshold 0.4
[debug] [MJSONWP (f720090e)] Encountered internal error running command: Error: This driver does not support the required 'getWindowSize' command`

And previously working locators now fail:
`[debug] [MJSONWP (f720090e)] Calling AppiumDriver.findElement() with args: ["name","Open:","f720090e-64a9-4cd1-bd83-2fb3b983fcb5"]
[debug] [BaseDriver] Valid locator strategies for this request:
[debug] [MJSONWP (f720090e)] Encountered internal error running command: InvalidSelectorError: Locator Strategy 'name' is not supported for this session`

I tried the RegExs before the last commit you made, and thought it was worth noting that with https://github.com/mykola-mokhnach/appium-windows-driver/commit/f4989e6e3031f251e554f5b72d1619d7f656f98d commit reversed, the old behaviour returns: all locators other than image are fine, and image throws a "InvalidLocatorError"

**Comment by mykola-mokhnach** [comment]
Cool, this is already something. Let me do some more magic...

File: lib/commands/find.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/commands/find.js]: yes, this improvement could be done in a separate PR

**Comment by mykola-mokhnach** [comment]
After https://github.com/appium/appium-windows-driver/pull/67/commits/84570b7a045278b63a168fbdd5eab5db089542e6 the errors above should be gone ;)

**Comment by jsa34** [comment]
Fab - all locators are now functional!

Tried finding by image, but getting the following error (made sure it has started a driver session with a top level window handle defined):
org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: Cannot destructure property 'width' of '(intermediate value)' as it is undefined. (WARNING: The server did not provide any stacktrace information)

appium error:
[debug] [MJSONWP (d31fc9cf)] Calling AppiumDriver.findElement() with args: ["-image","***"]
[BaseDriver] Finding image element with match threshold 0.4
[debug] [WD Proxy] Proxying [GET /window/size] to [GET http://127.0.0.1:4724/wd/hub/session/E9BD5D69-A757-4EAA-8A55-A66A6C080A4C/window/size] with no body
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] ==========================================
[WinAppDriver] [STDOUT] GET /wd/hub/session/E9BD5D69-A757-4EAA-8A55-A66A6C080A4C/window/size HTTP/1.1
[WinAppDriver] [STDOUT] Accept: application/json, */*
[WinAppDriver] [STDOUT] Connection: close
[WinAppDriver] [STDOUT] Content-Type: application/json; charset=utf-8
[WinAppDriver] [STDOUT] Host: 127.0.0.1:4724
[WinAppDriver] [STDOUT] User-Agent: appium
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] HTTP/1.1 200 OK
[WinAppDriver] [STDOUT] Content-Length: 63
[WinAppDriver] [STDOUT] Content-Type: application/json
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] {"sessionId":"E9BD5D69-A757-4EAA-8A55-A66A6C080A4C","status":0}
[debug] [WD Proxy] Got response with status 200: {"sessionId":"E9BD5D69-A757-4EAA-8A55-A66A6C080A4C","status":0}
[debug] [MJSONWP (d31fc9cf)] Encountered internal error running command: TypeError: Cannot destructure property 'width' of '(intermediate value)' as it is undefined.
[debug] [MJSONWP (d31fc9cf)]     at WindowsDriver.findByImage (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\commands\find.js:221:17)
[debug] [MJSONWP (d31fc9cf)]     at runMicrotasks (<anonymous>)
[debug] [MJSONWP (d31fc9cf)]     at runNextTicks (internal/process/task_queues.js:62:5)
[debug] [MJSONWP (d31fc9cf)]     at processImmediate (internal/timers.js:429:9)
[debug] [MJSONWP (d31fc9cf)]     at WindowsDriver.findElement (***\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\commands\find.js:48:12)
[HTTP] <-- POST /wd/hub/session/d31fc9cf-761d-4e0d-a4a3-c1b198e323da/element 500 38 ms - 252


Example of driver creation code:
            DesiredCapabilities capabilities = new DesiredCapabilities();
            capabilities.setCapability("platformName", "Windows");
            capabilities.setCapability("deviceName", "WindowsPC");
            capabilities.setCapability("ms:experimental-webdriver", true);
            capabilities.setCapability("appTopLevelWindow", desiredWindowHandle);
            WindowsDriver<WindowsElement> d = new WindowsDriver<>(getDriver().getRemoteAddress(), capabilities);

**Comment by mykola-mokhnach** [comment]
> Also tried d.manage().window().getSize() and this returned a null pointer... Suspect this is the root cause

This API is expected to return `{width, height}` set to valid window dimension values. If case the returned value is `null` then it is a highly possible bug inside https://github.com/microsoft/WinAppDriver

**Comment by mykola-mokhnach** [comment]
Btw, they declare the `GET /window/size` endpoint as supported: https://github.com/microsoft/WinAppDriver/blob/master/Docs/SupportedAPIs.md

**Comment by jsa34** [comment]
Hello - I am using 1.2-RC... I can't see why this is returning as I too saw it listed in their supported endpoints...

**Comment by jsa34** [comment]
Is there a way of trying different winappdriver versions, as using 1.7.0 beta seems to tie me to 1.2-RC version so I can't verify if other versions of winappdriver support this endpoint?

**Comment by jsa34** [comment]
No worries - found how to do it... Just trying different winappdriver versions...

**Comment by jsa34** [comment]
So, tried the following, but no luck:

- switching to v1.1: same behaviour,
-making sure the driver.manage().switchTo().window(currentWindowHandle); is set results in same behaviour, too

I have filed an issue with WAD team, but they don't seem to maintaining it from recent activity...

**Comment by jsa34** [comment]
Just seen a possible workaround would be to use the element of the window instead of the /window/size endpoint as they should be equal according to the test:
https://github.com/microsoft/WinAppDriver/blob/master/Tests/WebDriverAPI/ElementSize.cs (line 55-58)

Not sure if this also returns null or if it is a viable workaround, but just thought it worth mentioning

**Comment by mykola-mokhnach** [comment]
I've added a workaround there. I don't like it, but that's anyway better than waiting until MS fixes the server. Please try it and let me know if it works

**Comment by jsa34** [comment]
Great!  This is working now!
However, appium throws the following:
```
[debug] [WD Proxy] Got response with status 200: {"sessionId":"B340B876-CEC8-4CF8-AF10-191D6D179A12","status":0}
[WinAppDriver] Cannot retrieve window size from WinAppDriver. Falling back to WMIC usage
[debug] [WinAppDriver]
[debug] [WinAppDriver]
[debug] [WinAppDriver] CurrentHorizontalResolution=1680
[debug] [WinAppDriver] CurrentVerticalResolution=1050
[debug] [WinAppDriver] VideoModeDescription=1680 x 1050 x 4294967296 colors
[debug] [WinAppDriver]
[debug] [WinAppDriver]
[debug] [WinAppDriver]
[debug] [WinAppDriver]
[debug] [BaseDriver] Waiting up to 0 ms for condition
[debug] [WD Proxy] Matched '/screenshot' to command name 'getScreenshot'
[debug] [WD Proxy] Proxying [GET /screenshot] to [GET http://127.0.0.1:4724/wd/hub/session/B340B876-CEC8-4CF8-AF10-191D6D179A12/screenshot] with no body
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] ==========================================
[WinAppDriver] [STDOUT] GET /wd/hub/session/B340B876-CEC8-4CF8-AF10-191D6D179A12/screenshot HTTP/1.1
[WinAppDriver] [STDOUT] Accept: application/json, */*
[WinAppDriver] [STDOUT] Connection: close
[WinAppDriver] [STDOUT] Content-Type: application/json; charset=utf-8
[WinAppDriver] [STDOUT] Host: 127.0.0.1:4724
[WinAppDriver] [STDOUT] User-Agent: appium
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT] HTTP/1.1 200 OK
[WinAppDriver] [STDOUT] Content-Length: 259414
[WinAppDriver] [STDOUT] Content-Type: application/json
[WinAppDriver] [STDOUT]
[WinAppDriver] [STDOUT]
[debug] [WD Proxy] Got response with status 200: {"sessionId":"B340B876-CEC8-4CF8-AF10-191D6D179A12","status":0,"value":"***...
[BaseDriver] Verifying screenshot size and aspect ratio
[BaseDriver] When trying to find an element, determined that the screen aspect ratio and screenshot aspect ratio are different. Screen is 1680x1050 whereas screenshot is 1920x1040.
[BaseDriver] Resizing screenshot to 1901.7142857142858x1030.095238095238 to match screen aspect ratio so that image element coordinates have a greater chance of being correct.
[BaseDriver] Scaling screenshot from 1902x1030 to match screen at 1680x1050
[BaseDriver] Image template matched: {"x":112,"y":324,"width":70,"height":18}
[debug] [MJSONWP (4eae3a18)] Responding to client with driver.findElement() result: {"element-6066-11e4-a52e-4f735466cecf":"appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd","ELEMENT":"appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd"}
[HTTP] <-- POST /wd/hub/session/4eae3a18-2660-4521-80b2-04a056de5f3c/element 200 7443 ms - 241
[HTTP]
[HTTP] --> GET /wd/hub/session/4eae3a18-2660-4521-80b2-04a056de5f3c/element/appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd/displayed
[HTTP] {}
[debug] [MJSONWP (4eae3a18)] Calling AppiumDriver.elementDisplayed() with args: ["appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd","4eae3a18-2660-4521-80b2-04a056de5f3c"]
[debug] [MJSONWP (4eae3a18)] Responding to client with driver.elementDisplayed() result: true
[HTTP] <-- GET /wd/hub/session/4eae3a18-2660-4521-80b2-04a056de5f3c/element/appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd/displayed 200 4 ms - 76
[HTTP]
[HTTP] --> GET /wd/hub/session/4eae3a18-2660-4521-80b2-04a056de5f3c/element/appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd/enabled
[HTTP] {}
[debug] [MJSONWP (4eae3a18)] Calling AppiumDriver.elementEnabled() with args: ["appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd","4eae3a18-2660-4521-80b2-04a056de5f3c"]
[debug] [MJSONWP (4eae3a18)] Encountered internal error running command: NotYetImplementedError: Method has not yet been implemented
[debug] [MJSONWP (4eae3a18)]     at Function.execute (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\image-element.js:250:22)
[debug] [MJSONWP (4eae3a18)]     at commandExecutor (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\driver.js:328:28)
[debug] [MJSONWP (4eae3a18)]     at C:\Users\automation\Documents\local_modules\appium\node_modules\async-lock\lib\index.js:125:12
[debug] [MJSONWP (4eae3a18)]     at AsyncLock._promiseTry (C:\Users\automation\Documents\local_modules\appium\node_modules\async-lock\lib\index.js:249:31)
[debug] [MJSONWP (4eae3a18)]     at exec (C:\Users\automation\Documents\local_modules\appium\node_modules\async-lock\lib\index.js:124:9)
[debug] [MJSONWP (4eae3a18)]     at AsyncLock.acquire (C:\Users\automation\Documents\local_modules\appium\node_modules\async-lock\lib\index.js:140:3)
[debug] [MJSONWP (4eae3a18)]     at WindowsDriver.executeCommand (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\driver.js:343:39)
[debug] [MJSONWP (4eae3a18)]     at AppiumDriver.executeCommand (C:\Users\automation\Documents\local_modules\appium\lib\appium.js:534:36)
[debug] [MJSONWP (4eae3a18)]     at processTicksAndRejections (internal/process/task_queues.js:97:5)
[debug] [MJSONWP (4eae3a18)]     at asyncHandler (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\protocol\protocol.js:309:21)
[HTTP] <-- GET /wd/hub/session/4eae3a18-2660-4521-80b2-04a056de5f3c/element/appium-image-element-279a312f-8946-40a9-b80f-ee410465d4fd/enabled 501 74 ms - 123
```

**Comment by jsa34** [comment]
Just ran it a second time and got this strange error...

```
[debug] [WD Proxy] Got response with status 200: {"sessionId":"79B06E6E-B32F-4954-A204-1078C903C4B6","status":0,"value":"***...
[BaseDriver] Verifying screenshot size and aspect ratio
[BaseDriver] When trying to find an element, determined that the screen aspect ratio and screenshot aspect ratio are different. Screen is 1680x1050 whereas screenshot is 1920x1040.
[BaseDriver] Resizing screenshot to 1901.7142857142858x1030.095238095238 to match screen aspect ratio so that image element coordinates have a greater chance of being correct.
[BaseDriver] Scaling screenshot from 1902x1030 to match screen at 1680x1050
[BaseDriver] Image template matched: {"x":112,"y":324,"width":70,"height":18}
[BaseDriver] Will tap on image element at coordinate [147, 333]
[BaseDriver] Will tap using W3C actions
[BaseDriver] Driver does not seem to implement W3C actions, falling back to TouchActions
[BaseDriver] Will tap using MJSONWP TouchActions
[debug] [MJSONWP (d5e83d88)] Encountered internal error running command: Error: Driver did not implement the 'performTouch' command. For drivers to support finding image elements, they should support 'performTouch' and 'performActions'
[debug] [MJSONWP (d5e83d88)]     at ImageElement.click (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\image-element.js:207:11)
[debug] [MJSONWP (d5e83d88)]     at Function.execute (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\image-element.js:231:16)
[debug] [MJSONWP (d5e83d88)]     at commandExecutor (C:\Users\automation\Documents\local_modules\appium\node_modules\appium-base-driver\lib\basedriver\driver.js:328:9)
[HTTP] <-- POST /wd/hub/session/d5e83d88-dab1-4882-8685-44a6818a025b/element/appium-image-element-15224153-153e-40a5-a3d8-e3df69d86e9d/click 500 1068 ms - 326
```

**Comment by mykola-mokhnach** [comment]
These errors are expected. We'd have to add a wrapper for touch actions to make clicks working same way the wrappers for screenshots/screensize have been added, but not in this PR.
Would you like to try it yourself?

File: lib/commands/general.js [hunk 2]
```diff
@@ -1,9 +1,34 @@
...
 // for proper `-image` locator functionality
 commands.getWindowSize = async function getWindowSize () {
-  return await this.winAppDriver.sendCommand('/window/size', 'GET');
+  const size = await this.winAppDriver.sendCommand('/window/size', 'GET');
+  if (_.isPlainObject(size)) {
+    return size;
+  }
+  // workaround for https://github.com/microsoft/WinAppDriver/issues/1104
+  log.info('Cannot retrieve window size from WinAppDriver. ' +
+    'Falling back to WMIC usage');
```
Comment (KazuCocoa) [inline lib/commands/general.js]: wow...ah..

**Comment by jsa34** [comment]
Thanks for the explanation - I will have a look at the wrapper solution for touch actions, but I still don't understand why: 
`WindowsElement e = driver.findElementByImage(ImageManager.getImage(imageName));
`Where driver is definitely of type WindowsDriver<WindowsElement> throws the following error:
`java.lang.ClassCastException: org.openqa.selenium.remote.RemoteWebElement cannot be cast to io.appium.java_client.windows.WindowsElement
`
It's something internal where .findElementByImage should return back a WindowsElement but it looks like some internal casting to RemoteWebElement causes issues with returning this type back?

**Pull request merged**